python/ray/autoscaler/aliyun/example-linux.yaml

```diff
@@ -0,0 +1,171 @@
+        # resources: {"CPU": 1, "GPU": 1, "custom": 5}
+        resources: {"CPU": 8}
+        # Provider-specific config for this node type, e.g. instance type. By default
+        # Ray will auto-configure unspecified fields such as SubnetId and KeyName.
+        # For more documentation on available fields, see:
+        # http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.ServiceResource.create_instances
+        node_config:
+            InstanceType: ecs.n4.2xlarge
+            ImageId: ubuntu_20_04_x64_20G_alibase_20210420.vhd
+#            KeyPairName: id_rsa.pub
+            # Run workers on spot by default. Comment this out to use on-demand.
+            InstanceMarketOptions:
+                MarketType: spot
+                # Additional options can be found in the boto docs, e.g.
+                #   SpotOptions:
+                #       MaxPrice: MAX_HOURLY_PRICE
+
+
+
+# Specify the node type of the head node (as configured above).
+head_node_type: ray.head.default
+
+# Files or directories to copy to the head and worker nodes. The format is a
+# dictionary from REMOTE_PATH: LOCAL_PATH, e.g.
+file_mounts: {
+#    "~/dist":"~/alipay/ray/python/dist",
+#    "/path2/on/remote/machine": "/path2/on/local/machine",
+}
+
+# Files or directories to copy from the head node to the worker nodes. The format is a
+# list of paths. The same path on the head node will be copied to the worker node.
+# This behavior is a subset of the file_mounts behavior. In the vast majority of cases
+# you should just use file_mounts. Only use this if you know what you're doing!
+cluster_synced_files: []
+
+# Whether changes to directories in file_mounts or cluster_synced_files in the head node
+# should sync to the worker node continuously
+file_mounts_sync_continuously: True
+
+# Patterns for files to exclude when running rsync up or rsync down
+rsync_exclude: []
+
+# Pattern files to use for filtering out files when running rsync up or rsync down. The file is searched for
+# in the source directory and recursively through all subdirectories. For example, if .gitignore is provided
+# as a value, the behavior will match git's behavior for finding and using .gitignore files.
+rsync_filter: []
+
+# List of commands that will be run before `setup_commands`. If docker is
+# enabled, these commands will run outside the container and before docker
+# is setup.
+initialization_commands: []
+
+# List of shell commands to run to set up nodes.
+setup_commands:
+    - sudo apt-get update
+    # Install Anaconda.
+    - wget https://repo.continuum.io/archive/Anaconda3-2020.11-Linux-x86_64.sh || true
+    - bash Anaconda3-2020.11-Linux-x86_64.sh -b -p $HOME/anaconda3 || true
+    - echo 'export PATH="$HOME/anaconda3/bin:$PATH"' >> ~/.bashrc
+    # Install Ray
+    - pip install pytest-runner
+    - pip install -U ray
+    # Install Aliyun skd
+    - pip install aliyun-python-sdk-core
+    - pip install aliyun-python-sdk-ecs
```

Comment:
Do these need to be on all nodes or just the head node? (This applies to the other YAML files as well).

Reply:
Just the head node needs these. I have moved that to `head_setup_commands`